cranelift-codegen/src/isa/x86/unwind.rs

```diff
@@ -0,0 +1,508 @@
+                )?;
+            }
+            Self::StackAlloc { offset, size } => {
+                // Stack allocations on Windows must be a multiple of 8 and be at least 1 slot
+                assert!(*size >= 8);
+                assert!((*size % 8) == 0);
+
+                mem.write_u8(*offset)?;
+                if *size <= SMALL_ALLOC_MAX_SIZE {
+                    mem.write_u8(
+                        ((((*size - 8) / 8) as u8) << 4) | UnwindOperation::SmallStackAlloc as u8,
+                    )?;
+                } else if *size <= LARGE_ALLOC_16BIT_MAX_SIZE {
+                    mem.write_u8(UnwindOperation::LargeStackAlloc as u8)?;
+                    mem.write_u16::<LittleEndian>((*size / 8) as u16)?;
+                } else {
+                    mem.write_u8((1 << 4) | (UnwindOperation::LargeStackAlloc as u8))?;
+                    mem.write_u32::<LittleEndian>(*size)?;
+                }
+            }
+            Self::SetFramePointer { offset, sp_offset } => {
+                mem.write_u8(*offset)?;
+                mem.write_u8((*sp_offset << 4) | (UnwindOperation::SetFramePointer as u8))?;
+            }
+        };
+
+        Ok(())
+    }
+
+    fn node_count(&self) -> usize {
+        match self {
+            Self::StackAlloc { size, .. } => {
+                if *size <= SMALL_ALLOC_MAX_SIZE {
+                    1
+                } else if *size <= LARGE_ALLOC_16BIT_MAX_SIZE {
+                    2
+                } else {
+                    3
+                }
+            }
+            _ => 1,
+        }
+    }
+}
+
+/// Represents Windows x64 unwind information.
+///
+/// For information about Windows x64 unwind info, see:
+/// https://docs.microsoft.com/en-us/cpp/build/exception-handling-x64
+#[derive(Debug, PartialEq, Eq)]
+pub struct UnwindInfo {
+    flags: u8,
+    prologue_size: u8,
+    frame_register: Option<RegUnit>,
+    frame_register_offset: u8,
+    unwind_codes: Vec<UnwindCode>,
+}
+
+impl UnwindInfo {
+    pub fn try_from_func(
+        func: &Function,
+        isa: &dyn TargetIsa,
+        frame_register: Option<RegUnit>,
+    ) -> Option<Self> {
+        // Only Windows fastcall is supported for unwind information
+        if func.signature.call_conv != CallConv::WindowsFastcall || func.prologue_end.is_none() {
+            return None;
+        }
+
+        let prologue_end = func.prologue_end.unwrap();
+        let entry_block = func.layout.ebbs().nth(0).expect("missing entry block");
+
+        // Stores the stack size when SP is not adjusted via an immediate value
+        let mut stack_size = None;
+        let mut prologue_size = 0;
+        let mut unwind_codes = Vec::new();
+        let mut found_end = false;
+
+        for (offset, inst, size) in func.inst_offsets(entry_block, &isa.encoding_info()) {
+            // x64 ABI prologues cannot exceed 256 bytes in length
+            if (offset + size) > 256 {
+                return None;
```

Comment:
If we ever do emit a prologue that big, what are the consequences if we try to unwind through such a function, without unwind information?

In any case, this seems like a good place for a `debug!` message.

Reply:
It would likely result in the operating system raising an unhandled invalid stack exception whenever it attempts to unwind through such a frame.  This is what's currently happening for most traps in Wasmtime on Windows.

It shouldn't be possible for us to emit a prologue that big for the Windows ABI, so I think we could perhaps panic/ICE if we hit this condition.

Reply:
Makes sense. A panic sounds reasonable here.

Reply:
I'll add.